go/kbfs/libkbfs/settings_db.go

```diff
@@ -0,0 +1,142 @@
+package libkbfs
+
+import (
+	"os"
+	"path"
+	"strconv"
+
+	"github.com/keybase/client/go/kbfs/idutil"
+	"github.com/keybase/client/go/protocol/keybase1"
+	"github.com/pkg/errors"
+	"github.com/syndtr/goleveldb/leveldb/opt"
+	"github.com/syndtr/goleveldb/leveldb/storage"
+	"golang.org/x/net/context"
+)
+
+const (
+	// Where in config.StorageRoot() we store settings information.
+	settingsDBDir           = "kbfs_settings"
+	settingsDBVersionString = "v1"
+	settingsDBName          = "kbfsSettings.leveldb"
+
+	// Settings keys
+	notificationThresholdKey = "notificationThreshold"
+)
+
+// ErrNoSettingsDB is returned when there is no settings DB potentially due to
+// multiple concurrent KBFS instances.
+var ErrNoSettingsDB = errors.New("no settings DB")
+
+var errNoSession = errors.New("no session")
+
+// SettingsDB stores KBFS user settings for a given device.
+type SettingsDB struct {
+	*LevelDb
+	sessionGetter idutil.CurrentSessionGetter
+
+	cache map[string][]byte
+}
+
+func openSettingsDBInternal(config Config) (*LevelDb, error) {
+	if config.IsTestMode() {
+		return openLevelDB(storage.NewMemStorage())
+	}
+	dbPath := path.Join(config.StorageRoot(), settingsDBDir,
+		settingsDBVersionString)
+	err := os.MkdirAll(dbPath, os.ModePerm)
+	if err != nil {
+		return nil, err
+	}
+
+	stor, err := storage.OpenFile(path.Join(dbPath, settingsDBName), false)
+	if err != nil {
+		return nil, err
+	}
+
+	return openLevelDB(stor)
+}
+
+func openSettingsDB(config Config) *SettingsDB {
+	db, err := openSettingsDBInternal(config)
+	if err != nil {
+		config.MakeLogger("").CWarningf(context.Background(),
+			"Could not open settings DB. "+
+				"Perhaps multiple KBFS instances are being run concurrently"+
+				"? Error: %+v", err)
+		if db != nil {
+			db.Close()
+		}
+		return nil
+	}
+	return &SettingsDB{
+		LevelDb:       db,
+		sessionGetter: config.CurrentSessionGetter(),
+	}
+}
+
+func (db *SettingsDB) getUID(ctx context.Context) keybase1.UID {
+	si, err := db.sessionGetter.GetCurrentSession(ctx)
+	if err != nil {
+		return keybase1.UID("")
+	}
+	return si.UID
+}
+
+func getSettingsDbKey(uid keybase1.UID, key string) []byte {
+	return append([]byte(uid), []byte(key)...)
+}
+
+// Get overrides (*LevelDb).Get to cache values in memory.
+func (db *SettingsDB) Get(key []byte, ro *opt.ReadOptions) ([]byte, error) {
```

Comment:
should `Put` and `Get` be exported? It seems like callers are not expected to use them.

Reply:
They are already exported, as per the comment. I'm just overriding them here.